components/style/properties/shorthand/background.mako.rs

```diff
@@ -12,172 +12,208 @@
+            % endfor
+
+            loop {
+                if background_color.is_none() {
+                    if let Ok(value) = input.try(|input| background_color::parse(context, input)) {
+                        background_color = Some(value);
+                        continue
+                    }
+                } else {
+                    // color can only be the last element
+                    return Err(())
                 }
-            }
-            if clip.is_none() {
-                if let Ok(value) = input.try(|input| background_clip::parse(context, input)) {
-                    clip = Some(value);
-                    any = true;
-                    continue
+                if position.is_none() {
+                    if let Ok(value) = input.try(|input| background_position::single_value::parse(context, input)) {
+                        position = Some(value);
+                        any_position = true;
+
+                        // Parse background size, if applicable.
+                        size = input.try(|input| {
+                            try!(input.expect_delim('/'));
+                            background_size::single_value::parse(context, input)
+                        }).ok();
+                        if let Some(_) = size {
+                            any_size = true;
+                        }
+
+                        continue
+                    }
                 }
+                % for name in "image repeat attachment origin clip".split():
+                    if ${name}.is_none() {
+                        if let Ok(value) = input.try(|input| background_${name}::single_value::parse(context, input)) {
+                            ${name} = Some(value);
+                            any_${name} = true;
+                            continue
+                        }
+                    }
+                % endfor
+                break
             }
-            break
+            Ok((image, position, repeat, size, attachment, origin, clip))
+        }));
+
+        if !(any_image || any_position || any_repeat || any_size ||
+             any_attachment || any_origin || any_clip ||
+             background_color.is_some()) {
+            return Err(())
         }
 
-        if any {
-            Ok(Longhands {
-                background_color: color,
-                background_image: image,
-                background_position: position,
-                background_repeat: repeat,
-                background_attachment: attachment,
-                background_size: size,
-                background_origin: origin,
-                background_clip: clip,
-            })
-        } else {
-            Err(())
+        % for name in "image position repeat size attachment origin clip".split():
+            let mut background_${name} = if any_${name} {
+                Some(background_${name}::SpecifiedValue(Vec::new()))
+            } else {
+                None
```

Comment:
Or rather, only use `Vec::new()` here and move `Some(background_${name}::SpecifiedValue(…))` to the `Longhands` struct literal.